### PR TITLE
init commit

### PR DESCRIPTION
**Pull request by j-windsor**

### CODE REVIEW

File: src/components/bannerCard.tsx [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+            this.props.onSave(banner);
+
+            this.setState({ dirty: false, loading: false });
+
+            this.isSaveEnabled();
+        } catch (ex) {
+            this.setState({ messageErrorText: "Unable to save" });
+        }
+
```
Comment (dherre12345) [inline src/components/bannerCard.tsx]: nit: Extra line

File: src/components/bannerCard.tsx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        try {
+            await VstsService.instance.deleteWebGlobalMessageBanner(this.props.banner);
+
+            this.props.onDelete();
+        } catch (ex) {
+            this.setState({ messageErrorText: "Unable to delete" });
+        }
+    }
+
+    private isSaveEnabled(): boolean {
```
Comment (dherre12345) [inline src/components/bannerCard.tsx]: minor, maybe consider a getter?

File: src/components/bannerCard.tsx [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            && !this.isMessageExpired();
+    }
+
+    private ensureMarkedDirty(): void {
+        if (this.state.dirty === false) {
+            this.setState({ dirty: true });
+        }
+    }
+
+    private isMessageExpired(): boolean {
```
Comment (dherre12345) [inline src/components/bannerCard.tsx]: minor, maybe consider a getter?

File: src/components/expiryPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            dateError: false,
+        };
+    }
+
+    public componentDidMount(): void {
+        this.setupFields();
+    }
+
+    public componentDidUpdate(prevProps: IExpiryPickerProps): void {
+        if (JSON.stringify(prevProps.expiryDate) !== JSON.stringify(this.props.expiryDate)) {
```
Comment (dherre12345) [inline src/components/expiryPicker.tsx]: Need to stringify here?  Maybe direct compare (or getTime() comapre)?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    "package:prod": "npm run package:vsts-extension --vssconfig=prod",
+    "publish:prod": "npm run publish:vsts-extension --vssconfig=prod",
+    "package:vsts-extension": "tfx extension create --manifests ./ado-manifests/azure-devops-extension-base.json --overrides-file ./ado-manifests/azure-devops-extension-%npm_config_vssconfig%.json --rev-version",
+    "publish:vsts-extension": "tfx extension publish --manifests ./ado-manifests/azure-devops-extension-base.json --overrides-file ./ado-manifests/azure-devops-extension-%npm_config_vssconfig%.json",
+    "dev": "webpack-dev-server --hot --progress --colors --content-base ./ --https --port 44405"
+  },
+  "dependencies": {
+    "azure-devops-extension-api": "~1.157.0",
+    "azure-devops-extension-sdk": "~2.0.10",
+    "azure-devops-ui": "~1.158.0",
```
Comment (wswiatek) [inline package.json]: Btw, there is version 1.159.0 now I think.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    "css-loader": "~3.0.0",
+    "file-loader": "~4.0.0",
+    "mini-css-extract-plugin": "^0.8.0",
+    "node-sass": "~4.12.0",
+    "optimize-css-assets-webpack-plugin": "^5.0.3",
+    "path": "^0.12.7",
+    "rimraf": "~2.6.2",
+    "sass-loader": "~7.1.0",
+    "style-loader": "~0.23.1",
+    "tfx-cli": "^0.6.3",
```
Comment (wswiatek) [inline package.json]: Can we change version of "tfx-cli" to ~0.7.10 ?

File: src/components/bannerCard.scss [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+.banner-card {
+    margin-bottom: 15px;
+}
+
+.banner-item {
+    display: flex;
+    flex-direction: column;
+    width: 100%;
+    padding: 5px;
```
Comment (wswiatek) [inline src/components/bannerCard.scss]: 4px :)
General feedback - can we use multiplications of 2, 4, etc. ?

File: src/components/bannerCard.tsx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+            expanded: false,
+        };
+    }
+
+    public componentDidMount(): void {
+        this.setupFields();
+    }
+
+    public componentDidUpdate(prevProps: IBannerCardProps): void {
+        if (JSON.stringify(prevProps.banner) !== JSON.stringify(this.props.banner)) {
```
Comment (wswiatek) [inline src/components/bannerCard.tsx]: I probably asked this before but should we worry about performance when comparing the diff using JSON.stringify?

File: src/components/bannerCard.tsx [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+                                    size={StatusSize.l}
+                                />
+                            </div>
+                        </Tooltip>
+                        <FormItem className="header-message">
+                            <TextField
+                                prefixIconProps={{ iconName: Level[this.state.level] }}
+                                value={this.state.message}
+                                onChange={(e, newValue) => {
+
```
Comment (wswiatek) [inline src/components/bannerCard.tsx]: Remove empty line 82.

File: src/components/bannerCard.tsx (hunk 8 above)
Comment (j-windsor) [reply src/components/bannerCard.tsx]: It is the fastest method I have found to compare objects. Various hashing algorithms are significantly worse in perf.

File: src/components/bannerCard.tsx (hunk 2 above)
Comment (wswiatek) [reply src/components/bannerCard.tsx]: +1

File: src/components/bannerCard.tsx (hunk 3 above)
Comment (wswiatek) [reply src/components/bannerCard.tsx]: +1

File: src/components/bannerCard.tsx [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+                ...Statuses.Failed,
+                text: this.errorText,
+            };
+        }
+
+        if (!this.state.dirty) {
+            if (this.state.expirationDate == null) {
+                return {
+                    ...Statuses.Success,
+                    text: "This message will be shown indefinitely.",
```
Comment (wswiatek) [inline src/components/bannerCard.tsx]: Nit: "This" or "The" ?

Comment (wswiatek) [reply src/components/bannerCard.tsx]: It's a general nit for this method (there are multiple strings ...) :)

File: src/components/expiryPicker.tsx (hunk 4 above)
Comment (wswiatek) [reply src/components/expiryPicker.tsx]: +1

File: src/components/expiryPicker.tsx [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+                            }
+                        }}
+                    />
+                </FormItem>
+                <FormItem label="Expiry time (local time)" error={this.state.expiryEnabled && this.state.dateError}>
+                    <TextField
+                        className="expiry-field"
+                        disabled={!this.state.expiryEnabled}
+                        value={this.dateStringObservable}
+                        placeholder="mm/dd/yyyy hh/mm"
```
Comment (wswiatek) [inline src/components/expiryPicker.tsx]: Consider "hh:mm" instead of "hh/mm"

File: src/components/expiryPicker.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+                            this.dateStringObservable.value = dateString;
+
+                            const newMoment = moment(dateString, "MM/DD/YYYY HH:mm", true);
+
+                            if (newMoment.isValid()) {
+                                if (newMoment.isBefore(moment.now())) {
+                                    // Date before now
+                                    this.setState({ dateError: true });
+                                    if (this.props.onParseError != null) {
+                                        this.props.onParseError("Message expires in the past");
```
Comment (wswiatek) [inline src/components/expiryPicker.tsx]: Nit: "Message" or "The message" ?

File: src/components/expiryPicker.tsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+                                        this.props.onChange(newMoment.toDate());
+                                    }
+
+                                    this.setState({ dateError: false });
+                                }
+                            } else {
+                                // Date Invalid
+                                this.setState({ dateError: true });
+                                if (this.props.onParseError != null) {
+                                    this.props.onParseError("Invalid expiry date");
```
Comment (wswiatek) [inline src/components/expiryPicker.tsx]: Nit: "Expiry" or "Expiration" ?

File: src/hub.tsx [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+        try {
+            const banners = await VstsService.instance.getWebGlobalMessageBanners();
+            this.setState({ banners, loading: false });
+        } catch (ex) {
+            this.setState({ loading: false, errorText: `There was an error loading the banners: ${ex.message}` });
+        }
+    }
+
+    public render(): JSX.Element {
+
```
Comment (wswiatek) [inline src/hub.tsx]: Remove empty line 55.

File: src/hub.tsx [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+                },
+            },
+        ];
+    }
+
+    private async deleteAllBanners(): Promise<void> {
+        try {
+            await VstsService.instance.deleteWebGlobalMessageBanners();
+
+            // then make it reflect in the ui.
```
Comment (wswiatek) [inline src/hub.tsx]: Nit: "Then", not "then"

File: src/hub.tsx [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+    private onAddBannerClicked(): void {
+        const { banners } = this.state;
+        banners.push(new GlobalMessageBanner());
+        this.setState({ banners });
+    }
+
+    private async onAboutClicked(): Promise<void> {
+        const dialogService = await SDK.getService<IHostPageLayoutService>(CommonServiceIds.HostPageLayoutService);
+        dialogService.openMessageDialog(
+            "Copyright Microsoft 2019",
```
Comment (wswiatek) [inline src/hub.tsx]: I am not sure about putting year there ...

File: src/services/vsts-service.ts [hunk 17]
```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { CommonServiceIds, ILocationService } from "azure-devops-extension-api";
+import * as SDK from "azure-devops-extension-sdk";
+
+import { GlobalMessageBanner, ObjectListWithCount, WebGlobalMessageBanner } from "../models/global-message-banners";
+
+export class VstsService {
```
Comment (wswiatek) [inline src/services/vsts-service.ts]: Rename to AdoService (rename the file name as well).

File: src/components/bannerCard.tsx [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+    private async saveBanner(): Promise<void> {
+        this.setState({ loading: true });
+
+        const banner = this.props.banner;
+        banner.message = this.state.message;
+        banner.level = this.state.level;
+        banner.expirationDate = this.state.expirationDate;
+
+        try {
+            await VstsService.instance.saveWebGlobalMessageBanner(banner);
```
Comment (wswiatek) [inline src/components/bannerCard.tsx]: Can we drop using ".instance" somehow ?

File: src/components/bannerCard.tsx [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+        return (
+            <Card className="banner-card">
+                <div className="banner-item">
+                    <div className="banner-header">
+                        <Tooltip text={statusProps.text}>
+                            <div>
+                                <Status
+                                    {...statusProps}
+                                    text={this.errorText}
+                                    size={StatusSize.l}
```
Comment (wswiatek) [inline src/components/bannerCard.tsx]: Is "StatusSize.l" correct value?

Comment (j-windsor) [reply src/components/bannerCard.tsx]: It is! l for large

File: src/components/bannerCard.tsx (hunk 10 above)
Comment (j-windsor) [reply src/components/bannerCard.tsx]: Agreed. "The"

File: src/components/expiryPicker.tsx (hunk 11 above)
Comment (j-windsor) [reply src/components/expiryPicker.tsx]: good catch!

File: src/hub.tsx (hunk 16 above)
Comment (j-windsor) [reply src/hub.tsx]: We can use date to get the current year! Btw, In the next PR I am going to put github information in here.

File: src/components/bannerCard.tsx (hunk 18 above)
Comment (j-windsor) [reply src/components/bannerCard.tsx]: I am using export default. Let me know what you think.

File: package.json [hunk 20]
```diff
@@ -43,7 +43,7 @@
     "rimraf": "~2.6.2",
     "sass-loader": "~7.1.0",
     "style-loader": "~0.23.1",
-    "tfx-cli": "^0.6.3",
+    "tfx-cli": "^0.7.10",
```
Comment (wswiatek) [inline package.json]: ~0.7.10 (not ^0.7.10)

File: src/hub.tsx [hunk 21]
```diff
@@ -171,7 +170,7 @@ class HubComponent extends React.Component<{}, IHubComponentState> {
     private async onAboutClicked(): Promise<void> {
         const dialogService = await SDK.getService<IHostPageLayoutService>(CommonServiceIds.HostPageLayoutService);
         dialogService.openMessageDialog(
-            "Copyright Microsoft 2019",
+            `Copyright Microsoft ${new Date().getFullYear()}`,
```
Comment (wswiatek) [inline src/hub.tsx]: Nice!

**Pull request merged**